From 14c9171beeeb73d298960ed373d8a29aa43241c4 Mon Sep 17 00:00:00 2001
From: Misael Lopez Cruz <misael.lopez@ti.com>
Date: Fri, 17 Dec 2010 18:18:49 -0600
Subject: ASoC: ABE DSP: Allow flexible ping pong buffer size

Add flexibility to change ping-pong buffer size (upto max
size: 48KB). It's recommended to use ping-pong buffer in
steps of 1kB, a constraint has been added for that purpose.

Change-Id: I4efa1f634fa650587314b9e725695208249b6705
Signed-off-by: Misael Lopez Cruz <misael.lopez@ti.com>
Signed-off-by: Sebastien Guiriec <s-guiriec@ti.com>
Signed-off-by: Chris Kelly <c-kelly@ti.com>
---
 sound/soc/omap/omap-abe-dsp.c |   18 ++++++++++++------
 sound/soc/omap/omap-abe-dsp.h |    1 -
 2 files changed, 12 insertions(+), 7 deletions(-)

Index: Kernel/sound/soc/omap/omap-abe-dsp.c
===================================================================
--- Kernel.orig/sound/soc/omap/omap-abe-dsp.c	2011-01-19 00:14:51.053321509 +0530
+++ Kernel/sound/soc/omap/omap-abe-dsp.c	2011-01-19 00:14:54.873323913 +0530
@@ -98,8 +98,8 @@
 				  SNDRV_PCM_INFO_RESUME,
 	.formats		= SNDRV_PCM_FMTBIT_S16_LE |
 				  SNDRV_PCM_FMTBIT_S32_LE,
-	.period_bytes_min	= 24 * 1024,
-	.period_bytes_max	= 48 * 1024,
+	.period_bytes_min	= 4 * 1024,
+	.period_bytes_max	= 24 * 1024,
 	.periods_min		= 2,
 	.periods_max		= 2,
 	.buffer_bytes_max	= 24 * 1024 * 2,
@@ -1698,6 +1698,7 @@
 	struct snd_soc_dai *dai = rtd->cpu_dai;
 	/* TODO: do not use abe global structure to assign pdev */
 	struct platform_device *pdev = abe->pdev;
+	int ret = 0;
 
 	mutex_lock(&abe->mutex);
 
@@ -1714,19 +1715,22 @@
 		break;
 	case ABE_FRONTEND_DAI_LP_MEDIA:
 		snd_soc_set_runtime_hwparams(substream, &omap_abe_hardware);
+		ret = snd_pcm_hw_constraint_step(substream->runtime, 0,
+					 SNDRV_PCM_HW_PARAM_BUFFER_BYTES, 1024);
 		break;
 	default:
 		break;
 	}
 
 	mutex_unlock(&abe->mutex);
-	return 0;
+	return ret;
 }
 static int abe_ping_pong_init(struct snd_pcm_hw_params *params,
 	struct snd_pcm_substream *substream)
 {
 	struct snd_pcm_runtime *runtime = substream->runtime;
 	abe_data_format_t format;
+	size_t period_size;
 	u32 dst;
 
 	/*Storing substream pointer for irq*/
@@ -1738,6 +1742,8 @@
 	if (format.f == 44100)
 		abe_write_event_generator(EVENT_44100);
 
+	period_size = params_period_bytes(params);
+
 	/*Adding ping pong buffer subroutine*/
 	abe_add_subroutine(&abe_irq_pingpong_player_id,
 				(abe_subroutine2) abe_irq_pingpong_subroutine,
@@ -1746,17 +1752,17 @@
 	/* Connect a Ping-Pong cache-flush protocol to MM_DL port */
 	abe_connect_irq_ping_pong_port(MM_DL_PORT, &format,
 				abe_irq_pingpong_player_id,
-				N_SAMPLES_BYTES, &dst,
+				period_size, &dst,
 				PING_PONG_WITH_MCU_IRQ);
 
 	/* Memory mapping for hw params */
 	runtime->dma_area  = abe->io_base + ABE_DMEM_BASE_OFFSET_MPU +
 				ABE_VM_AESS_OFFSET + dst;
 	runtime->dma_addr  = 0;
-	runtime->dma_bytes = N_SAMPLES_BYTES * 2;
+	runtime->dma_bytes = period_size * 2;
 
 	/* Need to set the first buffer in order to get interrupt */
-	abe_set_ping_pong_buffer(MM_DL_PORT, N_SAMPLES_BYTES);
+	abe_set_ping_pong_buffer(MM_DL_PORT, period_size);
 
 	return 0;
 }
Index: Kernel/sound/soc/omap/omap-abe-dsp.h
===================================================================
--- Kernel.orig/sound/soc/omap/omap-abe-dsp.h	2011-01-14 00:04:13.474343378 +0530
+++ Kernel/sound/soc/omap/omap-abe-dsp.h	2011-01-19 00:14:54.873323913 +0530
@@ -24,7 +24,6 @@
 #ifndef __OMAP_ABE_DSP_H__
 #define __OMAP_ABE_DSP_H__
 
-#define N_SAMPLES_BYTES			(6144 * 4)	/* Samples * bytes */
 #define ABE_DMEM_BASE_OFFSET_MPU	0x80000		/* Ping start offset of DMEM */
 #define ABE_DMEM_BASE_ADDRESS_MPU	0x49080000L	/* Pong start offset of DMEM */
 #define ABE_DMEM_BASE_OFFSET_PING_PONG	0x4000		/* Ping pong buffer DMEM offset */
